reference/performance: add a use case for SPM evolve explanation (#2327)
pingcap/docs: reference/performance/execution-plan-bind.md

The default value of `tidb_evolve_plan_baselines` is `off`.

> **Note:**
>

> > The global variable doesn't take effect in the current session. It only takes effect in a newly created session. To enable automatic binding evolution in the current session, change the key word `global` to `session`.

Reviewer: ```suggestion
> The global variable does not take effect in the current session. It only takes effect in a newly created session. To enable automatic binding evolution in the current session, change the key word `global` to `session`.
```

> After the automatic binding evolution feature is enabled, if the optimal execution plan selected by the optimizer is not among the bound execution plans, the optimizer marks the plan as an execution plan that waits for verification. Every `bind-info-lease` (the default value is `3s`), an execution plan to be verified is selected and compared with a bound execution plan with the least cost, in terms of the actual execution time. If the plan to be verified is better, it is marked as a usable binding. The following use case describes the preceding process.

Reviewer: ```suggestion
After the automatic binding evolution feature is enabled, if the optimal execution plan selected by the optimizer is not among the bound execution plans, the optimizer marks the plan as an execution plan that waits for verification. Every `bind-info-lease` (the default value is `3s`) interval, an execution plan to be verified is selected and compared with the bound execution plan with the least cost in terms of the actual execution time. If the plan to be verified performs better in the execution time, this plan is marked as a usable binding. The following use case describes the above process.
```

> Let table `t` be defined as:

Reviewer: ```suggestion
Assume that table `t` is defined as follows:
```

{{< copyable "sql" >}}

```sql
create table t(a int, b int, key(a), key(b));
```

Perform the following query on table `t`:

{{< copyable "sql" >}}

```sql
select * from t where a < 100 and b < 100;
```

> Rows in the table that satisfy condition `a < 100` are rare. However, for some reasons, the optimizer mistakenly chooses the full table scan instead of the optimal execution plan that uses index `a`. Users can use the following statement to create a binding first:

Reviewer: ```suggestion
In the table defined above, few rows meet the `a < 100` condition. However, for some reason, the optimizer chooses the full table scan instead of the optimal execution plan that uses index `a`. You can first use the following statement to create a binding:
```

{{< copyable "sql" >}}

```sql
create global binding for select * from t where a < 100 and b < 100 using select * from t use index(a) where a < 100 and b < 100;
```

> When the previous query is executed again, the optimizer chooses index `a` under the interference of the binding created previously to reduce the query time.

Reviewer: ```suggestion
When the above query is executed again, the optimizer chooses index `a` under the interference of the binding created above to reduce the query time.
```

> Assuming that as insertions and deletions are performed on the table, the number of rows that satisfy condition `a < 100` grows while the number of rows that satisfy condition `b < 100` not, still using index `a` under the binding's interference might no longer be optimal.

Reviewer: ```suggestion
Assuming that as insertions and deletions are performed on table `t`, an increasing number of rows meet the `a < 100` condition and a decreasing number of rows meet the `b < 100` condition. At this time, using index `a` under the binding's interference might no longer be the optimal plan.
```

> The binding evolution can address this kind of problems. When the optimizer senses changes in table data, it generates an execution plan against the query that uses index `b`. However, because of the binding's presence, this query plan is not adopted and executed. Instead, it is stored in the backend evolution lists. During the evolution process, if it is verified that the time spent on executing the query plan is significantly shorter than that using index `a` (that is the current binding's execution plan), index `b` is added into available binding lists. After this process, when the query is executed again, the optimizer first generates the execution plan that uses index `b` and makes sure that it is in the binding lists. Then the optimizer adopts and executes it to reduce the query time after data updates.

Reviewer: ```suggestion
The binding evolution can address this kind of issues. When the optimizer recognizes data change in a table, it generates an execution plan for the query that uses index `b`. However, because the binding of the current plan exists, this query plan is not adopted and executed. Instead, this plan is stored in the backend evolution list. During the evolution process, if this plan is verified to have an obviously shorter execution time than that of the current execution plan that uses index `a`, index `b` is added into the available binding list. After this process, when the query is executed again, the optimizer first generates the execution plan that uses index `b` and makes sure that this plan is in the binding list. Then the optimizer adopts and executes this plan to reduce the query time after data updates.
```
